evm/src/prover.rs

```diff
@@ -499,29 +498,16 @@ where
     // When opening the `Z`s polys at the "next" point, need to look at the point `next_step` steps away.
     let next_step = 1 << quotient_degree_bits;
 
-    let powers_vec = SmallPowers::<F>::new(F::coset_shift().to_noncanonical_u64() as u32)
-        .take(degree << quotient_degree_bits)
-        .collect_vec();
-    // Evaluation of the first Lagrange polynomial on the LDE domain, with custom lde_onto_coset.
-    let lagrange_first = PolynomialValues::<F>::selector(degree, 0);
-    let coeffs = ifft(lagrange_first).lde(quotient_degree_bits);
-    let modified_poly: PolynomialCoeffs<F> = powers_vec
-        .iter()
-        .zip(&coeffs.coeffs)
-        .map(|(&r, &c)| r * c)
-        .collect::<Vec<_>>()
-        .into();
-    let lagrange_first = modified_poly.fft_with_options(None, None);
-    // Evaluation of the last Lagrange polynomial on the LDE domain, with custom lde_onto_coset.
-    let lagrange_last = PolynomialValues::<F>::selector(degree, degree - 1);
-    let coeffs = ifft(lagrange_last).lde(quotient_degree_bits);
-    let modified_poly: PolynomialCoeffs<F> = powers_vec
-        .into_iter()
-        .zip(&coeffs.coeffs)
-        .map(|(r, &c)| r * c)
-        .collect::<Vec<_>>()
-        .into();
-    let lagrange_last = modified_poly.fft_with_options(None, None);
+    // Evaluation of the first Lagrange polynomial on the LDE domain.
+    let lagrange_first = PolynomialValues::selector(degree, 0).lde_onto_coset(
+        quotient_degree_bits,
+        SmallPowers::new(F::coset_shift().to_noncanonical_u64() as u32),
+    );
+    // Evaluation of the last Lagrange polynomial on the LDE domain.
+    let lagrange_last = PolynomialValues::selector(degree, degree - 1).lde_onto_coset(
+        quotient_degree_bits,
+        SmallPowers::new(F::coset_shift().to_noncanonical_u64() as u32),
+    );
```

Comment:
Hmm, maybe I misunderstood, but wasn't one of the main goals of your original change to reuse the `SmallPowers` vector `power_vec`? Here it seems to be recomputed...

Reply:
Oops sorry, I had reverted it to have a fairer comparison with the main branch and didn't revert it back. Will update (though pending the overall need to have an actual `mul_u32` method).